### PR TITLE
Add WebSocketNetworkTransport.closeConnection

### PR DESCRIPTION
**Pull request by BoD**
Related to #3666

### CODE REVIEW

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt [hunk 1]
```diff
@@ -286,6 +295,18 @@ private constructor(
...
 
+  /**
+   * Close the connection to the server (if it's open).
+   *
+   * This can be used to force a reconnection to the server, for instance when new auth tokens should be passed to the headers.
+   *
+   * The given [reason] will be propagated to [Builder.reopenWhen] to determine if the connection should be reopened. If not, it will be
+   * propagated to any Flows waiting for responses.
+   */
+  fun closeConnection(reason: Throwable) {
```
Comment (BoD) [inline apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:306]: Chose this name because `reconnect` is not exactly what this does - but no strong opinion on it!

File: tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt [hunk 2]
```diff
@@ -216,4 +217,44 @@ class WebSocketErrorsTest {
...
+          true
+        }
+        .build()
+
+    apolloClient.subscription(CountSubscription(2, 100))
+        .toFlow()
+        .test {
+          awaitItem() // 0
+
+          (apolloClient.subscriptionNetworkTransport as WebSocketNetworkTransport).closeConnection(MyWebSocketReconnectException())
```
Comment (BoD) [inline tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt:248]: Not great - we could have an extension on `NetworkTransport` that checks the type to make it a bit nicer to use?

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:306]: Yep, you're right 👍

File: tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt (hunk 2 above)
Comment (martinbonnin) [reply tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt:248]: Agreed

```kotlin
fun NetworkTransport.closeConnection()
```

?

File: tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt [hunk 3]
```diff
@@ -216,4 +217,44 @@ class WebSocketErrorsTest {
...
+
+          awaitItem() // 0 again since we've re-subscribed
+          awaitItem() // 1
+          awaitComplete()
+        }
+
+    apolloClient.dispose()
+
+    // connectionInitCount is 2 since we returned true in webSocketReopenWhen
+    assertEquals(2, connectionInitCount)
```
Comment (martinbonnin) [inline tests/websockets/src/jvmTest/kotlin/WebSocketErrorsTest.kt:258]: nice test 👍

**Comment by martinbonnin** [comment]
Reminder to put something in the migration guide for users looking for `.reconnect()`

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt [hunk 4]
```diff
@@ -166,6 +165,12 @@ private constructor(
             return
           }
 
+          if (message is Close) {
+            messages.trySend(NetworkError(message.reason))
+            closeProtocol()
```
Comment (martinbonnin) [inline apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt:170]: Do you need to call `closeProtocol()` here given that it'll be closed when handling the `NetworkError` message.

2nd question: would it be worth reusing the `NetworkError` and save a queue/dequeue? Not that it is that important but it feels like the semantics are the same?

**Review by martinbonnin — CHANGES_REQUESTED** [review]
A few questions + missing extension function and 3.3.0 documentation updates

File: apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt (hunk 4 above)
Comment (BoD) [reply apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/network/ws/WebSocketNetworkTransport.kt]: Good point! `NetworkError` was enough, no need for specific command 👍

File: docs/source/migration/3.0.mdx [hunk 5]
```diff
@@ -725,6 +725,11 @@ Apollo Android 3.x instead uses `WebsocketNetworkTransport` to expose a simplifi
...
-
-If you feel some API is missing from `SubscriptionManager`, please [reach out](https://github.com/apollographql/apollo-kotlin/issues/new?assignees=&labels=%3Asparkles%3A+Type%3A+Feature&template=feature_request.md&title=) so we can discuss the best way to add them.
+If you were previously using `apolloClient.subscriptionManager.reconnect()` to force a Websocket reconnect (for example
+to send new authentication parameters), you can now
+call `apolloClient.subscriptionNetworkTransport.closeConnection(<your exception>)`: this will behave as if a network
+error occurred, and you can react to this by reconnecting with `reopenWhen`.
+
+If you feel some API is missing from `SubscriptionManager`,
+please [reach out](https://github.com/apollographql/apollo-kotlin/issues/new?assignees=&labels=%3Asparkles%3A+Type%3A+Feature&template=feature_request.md&title=)
+so we can discuss the best way to add them.
```
Comment (martinbonnin) [inline docs/source/migration/3.0.mdx:735]: This has to be in a separate PR so we can merge them separately

Comment (BoD) [reply docs/source/migration/3.0.mdx:735]: Woops you're right of course!

**Review by martinbonnin — APPROVED** [review]

**Comment by BoD** [comment]
The PR adding `closeConnection` has just been merged, so it will be available in the next release, in the meantime it can be tried out with the [snapshots](https://github.com/apollographql/apollo-kotlin#snapshots).
